### PR TITLE
fix for issue #57

### PR DESCRIPTION
**Pull request by sanjusoftware**
@NathanSweet as discussed with you [here](https://groups.google.com/forum/#!topic/yamlbeans-users/dZKp3LnttOM) here's the addition of allow/disallow duplicates feature along with tests supporting the same. Hope you can merge it and release a new version. I need to use this in my project :)

### CODE REVIEW

File: src/com/esotericsoftware/yamlbeans/YamlReader.java [hunk 1]
```diff
@@ -1,424 +1,435 @@
...
-			this(message, null);
-		}
-	}
-
-	public static void main (String[] args) throws Exception {
-		YamlReader reader = new YamlReader(new FileReader("test/test.yml"));
-		System.out.println(reader.read());
-	}
-}
+/*
```
Comment (NathanSweet) [inline src/com/esotericsoftware/yamlbeans/YamlReader.java]: Diff shows whitespace in entire file as changed. Could be a Git client setting.

File: src/com/esotericsoftware/yamlbeans/YamlReader.java [hunk 2]
```diff
@@ -1,424 +1,435 @@
...
+	private final YamlConfig config;
+	Parser parser;
+	boolean allowDuplicates = true;
+	private final Map<String, Object> anchors = new HashMap();
+
+	public YamlReader (Reader reader) {
+		this(reader, new YamlConfig());
+	}
+
+	public YamlReader (Reader reader, boolean allowDuplicates) {
```
Comment (NathanSweet) [inline src/com/esotericsoftware/yamlbeans/YamlReader.java]: Don't add a constructor, the setting should go in YamlConfig like all other settings.

File: src/com/esotericsoftware/yamlbeans/YamlReader.java [hunk 3]
```diff
@@ -1,424 +1,435 @@
...
+					key = nameValuePair.getKey();
+					value = nameValuePair.getValue();
+				}
+				if (object instanceof Map) {
+					// Add to map.
+					if (!isExplicitKey) value = readValue(elementType, null, null);
+					if (!allowDuplicates && keys.contains(key)) {
+						throw new YamlReaderException("Duplicate key found '" + key + "'");
+					}
+					keys.add(key);
```
Comment (NathanSweet) [inline src/com/esotericsoftware/yamlbeans/YamlReader.java]: `object` is a Map, just check if the map already has the value. Don't create an ArrayList when `object` is a Map. When it is NOT a map, then you need to use the `keys` ArrayList. Currently you don't find duplicate keys at all when not a map.

Comment (sanjusoftware) [inline src/com/esotericsoftware/yamlbeans/YamlReader.java]: agree.. I can just do 
```java
if (object instanceof Map) {
					// Add to map.
					if (!isExplicitKey) value = readValue(elementType, null, null);
					if (!config.allowDuplicates && ((Map) object).containsKey(key)) {
						throw new YamlReaderException("Duplicate key found '" + key + "'");
					}
					((Map)object).put(key, value);
				}
```
However, not entirely about where to add the NOT a map check.. can you help me with an example test yaml string that I can write a test for?

Comment (NathanSweet) [inline src/com/esotericsoftware/yamlbeans/YamlReader.java]: Yep, that looks good. When not  map, under `// Set field on object.` do your `keys.contains()` check.

**Comment by sanjusoftware** [comment]
done changed..raising another pull req

**Comment by sanjusoftware** [comment]
@NathanSweet let me know if this looks good to merge!!

**Comment by NathanSweet** [comment]
It looks good except the diff shows all the whitespace has changed. Can you fix that?

**Comment by sanjusoftware** [comment]
@NathanSweet yea I also noticed that.. not sure why is that happening... let me try and see what could case it.. any pointers will be helpful though :)

**Comment by NathanSweet** [comment]
Something about file endings, maybe:
https://help.github.com/articles/dealing-with-line-endings/

**Comment by sanjusoftware** [comment]
tried that with no luck. I am not sure where is this going wrong! on my local computer, I use JetBrains IntellijIdea IDE for development and when I do a diff there, it just shows the lines that I changed!

not sure where am I going wrong :(

**Comment by tomzo** [comment]
@sanjusoftware  This is a hexdump of file from your fork:
```console
$ hexdump -C src/com/esotericsoftware/yamlbeans/YamlReader.java  | head
00000000  2f 2a 0a 20 2a 20 43 6f  70 79 72 69 67 68 74 20  |/*. * Copyright |
00000010  28 63 29 20 32 30 30 38  20 4e 61 74 68 61 6e 20  |(c) 2008 Nathan |
00000020  53 77 65 65 74 0a 20 2a  20 0a 20 2a 20 50 65 72  |Sweet. * . * Per|
00000030  6d 69 73 73 69 6f 6e 20  69 73 20 68 65 72 65 62  |mission is hereb|
00000040  79 20 67 72 61 6e 74 65  64 2c 20 66 72 65 65 20  |y granted, free |
00000050  6f 66 20 63 68 61 72 67  65 2c 20 74 6f 20 61 6e  |of charge, to an|
00000060  79 20 70 65 72 73 6f 6e  20 6f 62 74 61 69 6e 69  |y person obtaini|
00000070  6e 67 20 61 20 63 6f 70  79 20 6f 66 20 74 68 69  |ng a copy of thi|
00000080  73 20 73 6f 66 74 77 61  72 65 20 61 6e 64 20 61  |s software and a|
00000090  73 73 6f 63 69 61 74 65  64 20 64 6f 63 75 6d 65  |ssociated docume|
```
And this is on master:
```console
hexdump -C src/com/esotericsoftware/yamlbeans/YamlReader.java  | head
00000000  2f 2a 0d 0a 20 2a 20 43  6f 70 79 72 69 67 68 74  |/*.. * Copyright|
00000010  20 28 63 29 20 32 30 30  38 20 4e 61 74 68 61 6e  | (c) 2008 Nathan|
00000020  20 53 77 65 65 74 0d 0a  20 2a 20 0d 0a 20 2a 20  | Sweet.. * .. * |
00000030  50 65 72 6d 69 73 73 69  6f 6e 20 69 73 20 68 65  |Permission is he|
00000040  72 65 62 79 20 67 72 61  6e 74 65 64 2c 20 66 72  |reby granted, fr|
00000050  65 65 20 6f 66 20 63 68  61 72 67 65 2c 20 74 6f  |ee of charge, to|
00000060  20 61 6e 79 20 70 65 72  73 6f 6e 20 6f 62 74 61  | any person obta|
00000070  69 6e 69 6e 67 20 61 20  63 6f 70 79 20 6f 66 20  |ining a copy of |
00000080  74 68 69 73 20 73 6f 66  74 77 61 72 65 20 61 6e  |this software an|
00000090  64 20 61 73 73 6f 63 69  61 74 65 64 20 64 6f 63  |d associated doc|
```

Notice that master has `0d` byte (which is a carriage return). You have removed that.

Probably original code was created on windows. It is up to @NathanSweet  to decide what style of line endings you want here. But that is only reason why git shows diff for full file.

**Comment by sanjusoftware** [comment]
@tomzo how do I fix it now though.. I followed the [link](https://help.github.com/articles/dealing-with-line-endings/) that @NathanSweet pointed me to, but its not helping anymore, as because I already did push the changes and now it has no effect and its keeping the line ending as LF (mac / unix style). I really want to get this change in!

**Comment by NathanSweet** [comment]
What are line endings for you locally? Replace \n with \r\n. Shouldn't need to commit a `.gitattributes` file.

**Comment by sanjusoftware** [comment]
@NathanSweet I am on a Mac machine... if possible can you help with me with the command please that I should run on project directory to replace \n with \r\n?

**Comment by NathanSweet** [comment]
Use Google! http://stackoverflow.com/questions/6373888/converting-newline-formatting-from-mac-to-windows

**Comment by sanjusoftware** [comment]
@NathanSweet sorry if that was annoying for you, but I really was trying to get it through.. I checked on my IDE once again and it says that all the files are with CRLF (windows style line separator). Can you confirm if this is now good to merge.

**Comment by tomzo** [comment]
>  I checked on my IDE once again and it says that all the files are with CRLF (windows style line separator). 

@sanjusoftware If you keep doing the same thing over and over again, it will not fix the issue. I showed you the wrong bytes that you have committed. The line endings are still unix style. You can check it with `hexdump -C`.

Running this on your modified files fixes the issue:
```
ex -bsc '%s/$/\n/|%!tr \\0 \\r' -cx  test/com/esotericsoftware/yamlbeans/YamlReaderTest.java
ex -bsc '%s/$/\n/|%!tr \\0 \\r' -cx  src/com/esotericsoftware/yamlbeans/YamlReader.java
ex -bsc '%s/$/\n/|%!tr \\0 \\r' -cx  src/com/esotericsoftware/yamlbeans/YamlConfig.java
```
See [this branch](https://github.com/tomzo/yamlbeans/tree/fix-endings) where I applied what you should have done. You can add my commit to your PR, but I would really prefer if you fixed this yourself.

**Comment by sanjusoftware** [comment]
@tomzo thanks so much for the help.. I now made the changes and pushed. 

PS: I looked at my command history as I realized I ran the command that you gave above already once, however, I found that I ran it on the whole repo root directory `ex -bsc '%s/$/\n/|%!tr \\0 \\r' -cx .` that time and that never returned for me somehow, which is when I aborted it.

@NathanSweet thanks for the awesome library and for bearing with me all this while. hope this is acceptable now.

**Comment by sanjusoftware** [comment]
@tomzo even if the above fixes the issue, I see that there one extra line added after each line of code somehow, which is looking very ugly!!

**Comment by sanjusoftware** [comment]
raised fresh PR already for this.. closing this one

**Pull request closed without merging**